### PR TITLE
Feature/tags grouping

### PR DESCRIPTION
**Pull request by michal1990**
In this PR exists one ignored test. It will be fixed in the next commit with a missing test.

### CODE REVIEW

File: app/src/main/java/com/invisibleteam/goinvisible/helper/EditItemAdapterHelper.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        advancedInfoTagList = new ArrayList<>();
+
+        tagsList = new ArrayList<>();
+    }
+
+    public void createGroups(List<Tag> tags) {
+        clearTagLists();
+        addHeadersToLists();
+        for (Tag tag : tags) {
+            if (IMAGE_INFO.equals(tag.getTagGroupType())) {
```
Comment (BartSoftware) [inline app/src/main/java/com/invisibleteam/goinvisible/helper/EditItemAdapterHelper.java]: switch would fit better

File: app/src/main/java/com/invisibleteam/goinvisible/model/Tag.java [hunk 2]
```diff
@@ -1,51 +1,37 @@
...
-        public Tag[] newArray(int size) {
-            return new Tag[size];
-        }
-    };
-
     private String key;
     private @Nullable String value;
     protected String formattedValue;
     private TagType tagType;
+    private TagGroupType tagGroupType;
```
Comment (BartSoftware) [inline app/src/main/java/com/invisibleteam/goinvisible/model/Tag.java:15]: @Nullable

File: app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapter.java [hunk 3]
```diff
@@ -1,64 +1,89 @@
...
+        if (position == IMAGE_INFO_SECTION_POSITION) {
+            return IMAGE_INFO.ordinal();
+        } else if (position == LOCATION_INFO_SECTION_POSITION) {
+            return LOCATION_INFO.ordinal();
+        } else if (position == DEVICE_INFO_SECTION_POSITION) {
+            return DEVICE_INFO.ordinal();
+        } else if (position == ADVANCED_INFO_SECTION_POSITION) {
+            return ADVANCED.ordinal();
+        } else {
+            return -1;
```
Comment (BartSoftware) [inline app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapter.java]: maybe better is to create some DEFAULT value

File: app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapter.java [hunk 4]
```diff
@@ -1,64 +1,89 @@
...
+    private static final int IMAGE_INFO_SECTION_POSITION = 0;
+    private static final int LOCATION_INFO_SECTION_POSITION = 21;
+    private static final int DEVICE_INFO_SECTION_POSITION = 47;
+    private static final int ADVANCED_INFO_SECTION_POSITION = 59;
 
-    private List<Tag> tagsList;
     private List<Tag> baseTagsList;
     private OnTagActionListener onTagActionListener;
+    private Context context;
+    private EditItemAdapterHelper helper;
```
Comment (BartSoftware) [inline app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapter.java]: final

File: app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapter.java [hunk 5]
```diff
@@ -1,64 +1,89 @@
...
+
+    private static final int IMAGE_INFO_SECTION_POSITION = 0;
+    private static final int LOCATION_INFO_SECTION_POSITION = 21;
+    private static final int DEVICE_INFO_SECTION_POSITION = 47;
+    private static final int ADVANCED_INFO_SECTION_POSITION = 59;
 
-    private List<Tag> tagsList;
     private List<Tag> baseTagsList;
     private OnTagActionListener onTagActionListener;
+    private Context context;
```
Comment (BartSoftware) [inline app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapter.java:33]: where is this context initialized?

File: app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapter.java [hunk 6]
```diff
@@ -1,64 +1,89 @@
...
-        });
-        viewHolder.itemView.setOnClickListener(view -> {
-            if (onTagActionListener != null) {
-                Tag tag = (Tag) itemView.getTag();
-                onTagActionListener.onEditStarted(tag);
-            }
-        });
+        context = parent.getContext();
+
+        if (viewType != -1) {
```
Comment (BartSoftware) [inline app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapter.java]: please make some defalut value, not hardcoded -1, it is used at least in 2 places

File: app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapter.java [hunk 7]
```diff
@@ -110,10 +132,11 @@ void setOnTagActionListener(OnTagActionListener listener) {
     }
 
     void updateTag(Tag tag) {
-        if (!tagsList.isEmpty()) {
-            for (int index = 0; index < tagsList.size(); index++) {
-                if (tagsList.get(index).getKey().equals(tag.getKey())) {
-                    tagsList.set(index, tag);
+        List<Tag> tagList = getTagsList();
+        if (!tagList.isEmpty()) {
```
Comment (BartSoftware) [inline app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapter.java]: if (tagList.isEmpty()) {
return
}

File: app/src/test/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapterTest.java [hunk 8]
```diff
@@ -60,16 +63,17 @@ public void init() {
...
         //Given
-        EditItemAdapter adapter = new EditItemAdapter();
+        EditItemAdapter adapter = new EditItemAdapter(new EditItemAdapterHelper());
         adapter.updateTagList(tagList);
         ViewGroup view = mock(ViewGroup.class);
         when(view.getContext()).thenReturn(context);
 
         //When
-        EditItemAdapter.ViewHolder holder = adapter.onCreateViewHolder(view, 0);
+        TagViewHolder holder = (TagViewHolder) adapter.onCreateViewHolder(view, -1);
```
Comment (BartSoftware) [inline app/src/test/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapterTest.java]: hardcoded -1

**Review by BartSoftware — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapter.java (hunk 3 above)
Comment (kejmil01) [reply app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapter.java]: Switch may be better

File: app/src/main/java/com/invisibleteam/goinvisible/model/Tag.java (hunk 2 above)
Comment (michal1990) [reply app/src/main/java/com/invisibleteam/goinvisible/model/Tag.java:15]: It's couldn't be nullable

File: app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapter.java (hunk 5 above)
Comment (michal1990) [reply app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapter.java:33]: In onCreateViewHolder

File: app/src/main/java/com/invisibleteam/goinvisible/helper/EditItemAdapterHelper.java (hunk 1 above)
Comment (michal1990) [reply app/src/main/java/com/invisibleteam/goinvisible/helper/EditItemAdapterHelper.java]: done

File: app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapter.java (hunk 4 above)
Comment (michal1990) [reply app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapter.java]: done

File: app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapter.java (hunk 6 above)
Comment (michal1990) [reply app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapter.java]: done

File: app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapter.java (hunk 3 above)
Comment (michal1990) [reply app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapter.java]: I made a switch with appropriate default

File: app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapter.java (hunk 7 above)
Comment (michal1990) [reply app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapter.java]: done

File: app/src/test/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapterTest.java (hunk 8 above)
Comment (michal1990) [reply app/src/test/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapterTest.java]: done

File: app/src/main/java/com/invisibleteam/goinvisible/model/Tag.java (hunk 2 above)
Comment (BartSoftware) [reply app/src/main/java/com/invisibleteam/goinvisible/model/Tag.java:15]: then there is non necessary condition while reading from parcelable

File: app/src/test/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapterTest.java [hunk 9]
```diff
@@ -1,96 +1,120 @@
...
     @Test
-    public void whenTagsListIsNotEmpty_ViewHolderIsBound() {
+    public void whenTagsListIsNotEmpty_TagViewHolderIsBound() {
         //Given
-        EditItemAdapter adapter = new EditItemAdapter();
+        EditItemAdapter adapter = new EditItemAdapter(new EditItemAdapterHelper());
         adapter.updateTagList(tagList);
-        ViewGroup view = mock(ViewGroup.class);
-        when(view.getContext()).thenReturn(context);
+        ViewGroup parent = new ViewGroup(activity) {
```
Comment (BartSoftware) [inline app/src/test/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/EditItemAdapterTest.java]: Better is to use some implementation like for example LinearLayout. It helps you avoid unnecessary methods implementation.

File: app/src/test/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/SectionViewHolderTest.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+@RunWith(RobolectricTestRunner.class)
+@Config(constants = BuildConfig.class, sdk = 21)
+public class SectionViewHolderTest {
+
+    private View itemView;
+
+    @Before
+    public void setUp() throws Exception {
+        EditActivity activity = Robolectric.buildActivity(EditActivity.class).create().get();
+        ViewGroup parent = new ViewGroup(activity) {
```
Comment (BartSoftware) [inline app/src/test/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/SectionViewHolderTest.java]: LinearLayout

File: app/src/test/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/TagViewHolderTest.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+@Config(constants = BuildConfig.class, sdk = 21)
+public class TagViewHolderTest {
+
+    private final Tag tag = new Tag("key", "value", TagType.build(InputType.TEXT_STRING), TagGroupType.IMAGE_INFO);
+    private View itemView;
+
+    @Before
+    public void setUp() throws Exception {
+        EditActivity activity = Robolectric.buildActivity(EditActivity.class).create().get();
+        ViewGroup parent = new ViewGroup(activity) {
```
Comment (BartSoftware) [inline app/src/test/java/com/invisibleteam/goinvisible/mvvm/edition/adapter/TagViewHolderTest.java]: LinearLayout

File: app/src/test/java/com/invisibleteam/goinvisible/mvvm/images/adapter/ImagesItemAdapterTest.java [hunk 12]
```diff
@@ -45,8 +46,12 @@ public void setUp() {
     @Test
     public void whenViewHolderIsCreatedWithNotEmptyList_AdapterHasResults() {
         //Given
-        ViewGroup parent = mock(ViewGroup.class);
-        when(parent.getContext()).thenReturn(activity);
+        ViewGroup parent = new ViewGroup(activity) {
```
Comment (BartSoftware) [inline app/src/test/java/com/invisibleteam/goinvisible/mvvm/images/adapter/ImagesItemAdapterTest.java]: LinearLayout

**Review by BartSoftware — CHANGES_REQUESTED** [review]

**Review by BartSoftware — APPROVED** [review]

**Pull request merged**